### PR TITLE
HL-15

### PR DESCRIPTION
**Pull request by OptimusOpus**
[Link to ticket](https://dltx.atlassian.net/jira/software/projects/HL/boards/82?selectedIssue=HL-15)

### CODE REVIEW

File: contracts/Vault.sol [hunk 1]
```diff
@@ -12,13 +12,20 @@ contract Vault is ERC4626Metadata, Ownable {
 
     // Mapping address => uint256
     mapping(address => uint256) public marketAllowance;
+
+    // Mapping address => uint256
+    // Unix time until receiver can withdraw from vault
+    mapping(address => uint256) public lockedTime;
```
Comment (brettvniekerk) [inline contracts/Vault.sol:18]: should we add a `getLockedTime` function so that users can see when they can withdraw their tokens?

**Review by bitcoinbrisbane — CHANGES_REQUESTED** [review]
Tests seem to be breaking

File: contracts/Vault.sol [hunk 2]
```diff
@@ -12,13 +12,20 @@ contract Vault is ERC4626Metadata, Ownable {
...
     
+    // Duration of lock up period in seconds
+    uint256 private _lockDuration;
```
Comment (Preston-Harrison) [inline contracts/Vault.sol:21]: It doesn't look like this changes. Could you make it immutable?

File: contracts/Vault.sol [hunk 3]
```diff
@@ -39,6 +47,10 @@ contract Vault is ERC4626Metadata, Ownable {
         return _decimals;
     }
 
+    function getLockDuration() public view returns (uint256) {
+        return _lockDuration;
```
Comment (Preston-Harrison) [inline contracts/Vault.sol:51]: why this, instead of making lockDuration public?

File: contracts/Vault.sol [hunk 4]
```diff
@@ -64,9 +76,21 @@ contract Vault is ERC4626Metadata, Ownable {
...
     function deposit(uint256 assets, address receiver) public override returns (uint256) {
         if (receiver == address(0)) receiver = _msgSender();
+        // Set lock time
+        lockedTime[receiver] = block.timestamp + _lockDuration;
         return super.deposit(assets, receiver);
     }
 
+    // If receiver is omitted, use the sender
+    function withdraw(uint256 assets, address receiver, address owner) public override returns (uint256) {
+        if (receiver == address(0)) receiver = _msgSender();
```
Comment (Preston-Harrison) [inline contracts/Vault.sol:86]: This is unusual. Instead, I'd create two functions, or an overloaded function.

**Review by Preston-Harrison — CHANGES_REQUESTED** [review]

File: contracts/Vault.sol (hunk 4 above)
Comment (Preston-Harrison) [inline contracts/Vault.sol:86]: Does this impose any checks on the caller? Seems like anyone can just call whatever

File: tests/vault.test.ts [hunk 5]
```diff
@@ -222,6 +237,67 @@ describe("Vault", () => {
...
+			const amount = ethers.utils.parseUnits("1000", underlyingDecimals);
+			await underlying.connect(alice).approve(vault.address, amount);
+
+			const latestBlockNumber = await ethers.provider.getBlockNumber();
+			const latestBlock = await ethers.provider.getBlock(latestBlockNumber);
+			await vault.connect(alice).deposit(amount, alice.address);
+
+			const lockedTime = await (
+				await vault.lockedTime(alice.address)
+			).toNumber();
```
Comment (Preston-Harrison) [inline tests/vault.test.ts:251]: the first await doesn't do anything here. For shorter syntax, you could do `+(await vault.lockedTime(alice.address)`

File: tests/vault.test.ts [hunk 6]
```diff
@@ -222,6 +237,67 @@ describe("Vault", () => {
...
+
+			expect(lockedTime).to.equal(
+				latestBlock.timestamp + Number(process.env.VAULT_LOCK_TIME) + 1
```
Comment (Preston-Harrison) [inline tests/vault.test.ts:254]: In the code styleguide, `+value` is preferred to `Number(value)`

File: tests/vault.test.ts [hunk 7]
```diff
@@ -222,6 +237,67 @@ describe("Vault", () => {
...
+						alice.address,
+						alice.address
+					)
+			).to.be.revertedWith("withdraw: Locked time not passed");
+
+			// Move past the lock up period
+			await hre.network.provider.request({
+				method: "evm_setNextBlockTimestamp",
+				params: [lockedTime + 1]
+			});
```
Comment (Preston-Harrison) [inline tests/vault.test.ts:270]: might be worth making this a util if there are going to be lots of time dependant tests

**Review by Preston-Harrison — CHANGES_REQUESTED** [review]
IMO these tests would be alot more readable without the giant nested await statements. Might be worth defining the awaited variables beforehand, then calling expect afterwards

File: contracts/Vault.sol (hunk 3 above)
Comment (bitcoinbrisbane) [reply contracts/Vault.sol:51]: I don't mind this pattern if you have the function as an interface function.

**Pull request closed without merging**